### ISSUE TITLE
ModuleNotFoundError: No module named 'torch_scatter.scatter_cuda' when test

### ISSUE THREAD

**Issue by HelloJocelynLu** [rusty1s/pytorch_scatter#62]
After install pytorch_scatter, it runs successfully under cpu environment:
`======================================================= test session starts ========================================================
platform linux -- Python 3.7.4, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
rootdir: /scratch/jojo/pytorch_scatter, inifile: setup.cfg
plugins: cov-2.7.1, remotedata-0.3.2, openfiles-0.3.2, arraydiff-0.3, doctestplus-0.3.0
collected 103 items                                                                                                                

test/test_backward.py .........
test/test_broadcasting.py .
test/test_forward.py ....................................................................................
test/test_max_min.py ..
test/test_multi_gpu.py s
test/test_std.py ......

----------- coverage: platform linux, python 3.7.4-final-0 -----------
Name                              Stmts   Miss  Cover
-----------------------------------------------------
torch_scatter/__init__.py            10      0   100%
torch_scatter/add.py                  4      0   100%
torch_scatter/div.py                 14      0   100%
torch_scatter/max.py                 19      0   100%
torch_scatter/mean.py                 6      0   100%
torch_scatter/min.py                 19      0   100%
torch_scatter/mul.py                 14      0   100%
torch_scatter/std.py                 16      0   100%
torch_scatter/sub.py                  3      0   100%
torch_scatter/utils/__init__.py       0      0   100%
torch_scatter/utils/ext.py            5      0   100%
torch_scatter/utils/gen.py           24      0   100%
-----------------------------------------------------
TOTAL                               134      0   100%

============================================== 102 passed, 1 skipped in 1.16 seconds ===============================================
`
But under GPU:
`ModuleNotFoundError: No module named 'torch_scatter.scatter_cuda'`

Extra information:
* System:
x86_64
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
* gcc --version: gcc (GCC) 6.3.0
* pytorch version: 1.2.0
* torch.version.cuda: 9.2.148
* Installation loh:
`running install
running bdist_egg
running egg_info
creating torch_scatter.egg-info
writing torch_scatter.egg-info/PKG-INFO
writing dependency_links to torch_scatter.egg-info/dependency_links.txt
writing top-level names to torch_scatter.egg-info/top_level.txt
writing manifest file 'torch_scatter.egg-info/SOURCES.txt'
reading manifest file 'torch_scatter.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'torch_scatter.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
creating build
creating build/lib.linux-x86_64-3.7
creating build/lib.linux-x86_64-3.7/torch_scatter
copying torch_scatter/max.py -> build/lib.linux-x86_64-3.7/torch_scatter
copying torch_scatter/__init__.py -> build/lib.linux-x86_64-3.7/torch_scatter
copying torch_scatter/std.py -> build/lib.linux-x86_64-3.7/torch_scatter
copying torch_scatter/min.py -> build/lib.linux-x86_64-3.7/torch_scatter
copying torch_scatter/mul.py -> build/lib.linux-x86_64-3.7/torch_scatter
copying torch_scatter/sub.py -> build/lib.linux-x86_64-3.7/torch_scatter
copying torch_scatter/add.py -> build/lib.linux-x86_64-3.7/torch_scatter
copying torch_scatter/div.py -> build/lib.linux-x86_64-3.7/torch_scatter
copying torch_scatter/mean.py -> build/lib.linux-x86_64-3.7/torch_scatter
creating build/lib.linux-x86_64-3.7/test
copying test/test_multi_gpu.py -> build/lib.linux-x86_64-3.7/test
copying test/test_max_min.py -> build/lib.linux-x86_64-3.7/test
copying test/__init__.py -> build/lib.linux-x86_64-3.7/test
copying test/test_std.py -> build/lib.linux-x86_64-3.7/test
copying test/utils.py -> build/lib.linux-x86_64-3.7/test
copying test/test_backward.py -> build/lib.linux-x86_64-3.7/test
copying test/test_broadcasting.py -> build/lib.linux-x86_64-3.7/test
copying test/test_forward.py -> build/lib.linux-x86_64-3.7/test
creating build/lib.linux-x86_64-3.7/torch_scatter/utils
copying torch_scatter/utils/__init__.py -> build/lib.linux-x86_64-3.7/torch_scatter/utils
copying torch_scatter/utils/ext.py -> build/lib.linux-x86_64-3.7/torch_scatter/utils
copying torch_scatter/utils/gen.py -> build/lib.linux-x86_64-3.7/torch_scatter/utils
running build_ext
building 'torch_scatter.scatter_cpu' extension
creating build/temp.linux-x86_64-3.7
creating build/temp.linux-x86_64-3.7/cpu
gcc -pthread -B /scratch/jojo/anaconda3/envs/torch-env/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/scratch/jojo/anaconda3/envs/torch-env/lib/python3.7/site-packages/torch/include -I/scratch/jojo/anaconda3/envs/torch-env/lib/python3.7/site-packages/torch/include/torch/csrc/api/include -I/scratch/jojo/anaconda3/envs/torch-env/lib/python3.7/site-packages/torch/include/TH -I/scratch/jojo/anaconda3/envs/torch-env/lib/python3.7/site-packages/torch/include/THC -I/scratch/jojo/anaconda3/envs/torch-env/include/python3.7m -c cpu/scatter.cpp -o build/temp.linux-x86_64-3.7/cpu/scatter.o -Wno-unused-variable -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=scatter_cpu -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++11
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
g++ -pthread -shared -B /scratch/jojo/anaconda3/envs/torch-env/compiler_compat -L/scratch/jojo/anaconda3/envs/torch-env/lib -Wl,-rpath=/scratch/jojo/anaconda3/envs/torch-env/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-x86_64-3.7/cpu/scatter.o -o build/lib.linux-x86_64-3.7/torch_scatter/scatter_cpu.cpython-37m-x86_64-linux-gnu.so
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/egg
creating build/bdist.linux-x86_64/egg/torch_scatter
copying build/lib.linux-x86_64-3.7/torch_scatter/max.py -> build/bdist.linux-x86_64/egg/torch_scatter
copying build/lib.linux-x86_64-3.7/torch_scatter/__init__.py -> build/bdist.linux-x86_64/egg/torch_scatter
copying build/lib.linux-x86_64-3.7/torch_scatter/std.py -> build/bdist.linux-x86_64/egg/torch_scatter
copying build/lib.linux-x86_64-3.7/torch_scatter/min.py -> build/bdist.linux-x86_64/egg/torch_scatter
copying build/lib.linux-x86_64-3.7/torch_scatter/mul.py -> build/bdist.linux-x86_64/egg/torch_scatter
copying build/lib.linux-x86_64-3.7/torch_scatter/sub.py -> build/bdist.linux-x86_64/egg/torch_scatter
creating build/bdist.linux-x86_64/egg/torch_scatter/utils
copying build/lib.linux-x86_64-3.7/torch_scatter/utils/__init__.py -> build/bdist.linux-x86_64/egg/torch_scatter/utils
copying build/lib.linux-x86_64-3.7/torch_scatter/utils/ext.py -> build/bdist.linux-x86_64/egg/torch_scatter/utils
copying build/lib.linux-x86_64-3.7/torch_scatter/utils/gen.py -> build/bdist.linux-x86_64/egg/torch_scatter/utils
copying build/lib.linux-x86_64-3.7/torch_scatter/add.py -> build/bdist.linux-x86_64/egg/torch_scatter
copying build/lib.linux-x86_64-3.7/torch_scatter/div.py -> build/bdist.linux-x86_64/egg/torch_scatter
copying build/lib.linux-x86_64-3.7/torch_scatter/scatter_cpu.cpython-37m-x86_64-linux-gnu.so -> build/bdist.linux-x86_64/egg/torch_scatter
copying build/lib.linux-x86_64-3.7/torch_scatter/mean.py -> build/bdist.linux-x86_64/egg/torch_scatter
creating build/bdist.linux-x86_64/egg/test
copying build/lib.linux-x86_64-3.7/test/test_multi_gpu.py -> build/bdist.linux-x86_64/egg/test
copying build/lib.linux-x86_64-3.7/test/test_max_min.py -> build/bdist.linux-x86_64/egg/test
copying build/lib.linux-x86_64-3.7/test/__init__.py -> build/bdist.linux-x86_64/egg/test
copying build/lib.linux-x86_64-3.7/test/test_std.py -> build/bdist.linux-x86_64/egg/test
copying build/lib.linux-x86_64-3.7/test/utils.py -> build/bdist.linux-x86_64/egg/test
copying build/lib.linux-x86_64-3.7/test/test_backward.py -> build/bdist.linux-x86_64/egg/test
copying build/lib.linux-x86_64-3.7/test/test_broadcasting.py -> build/bdist.linux-x86_64/egg/test
copying build/lib.linux-x86_64-3.7/test/test_forward.py -> build/bdist.linux-x86_64/egg/test
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/max.py to max.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/__init__.py to __init__.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/std.py to std.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/min.py to min.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/mul.py to mul.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/sub.py to sub.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/utils/__init__.py to __init__.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/utils/ext.py to ext.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/utils/gen.py to gen.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/add.py to add.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/div.py to div.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/mean.py to mean.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/test/test_multi_gpu.py to test_multi_gpu.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/test/test_max_min.py to test_max_min.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/test/__init__.py to __init__.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/test/test_std.py to test_std.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/test/utils.py to utils.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/test/test_backward.py to test_backward.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/test/test_broadcasting.py to test_broadcasting.cpython-37.pyc
byte-compiling build/bdist.linux-x86_64/egg/test/test_forward.py to test_forward.cpython-37.pyc
creating stub loader for torch_scatter/scatter_cpu.cpython-37m-x86_64-linux-gnu.so
byte-compiling build/bdist.linux-x86_64/egg/torch_scatter/scatter_cpu.py to scatter_cpu.cpython-37.pyc
creating build/bdist.linux-x86_64/egg/EGG-INFO
copying torch_scatter.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying torch_scatter.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying torch_scatter.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying torch_scatter.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
writing build/bdist.linux-x86_64/egg/EGG-INFO/native_libs.txt
zip_safe flag not set; analyzing archive contents...
torch_scatter.__pycache__.scatter_cpu.cpython-37: module references __file__
creating dist
creating 'dist/torch_scatter-1.3.1-py3.7-linux-x86_64.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing torch_scatter-1.3.1-py3.7-linux-x86_64.egg
creating /scratch/jojo/anaconda3/envs/torch-env/lib/python3.7/site-packages/torch_scatter-1.3.1-py3.7-linux-x86_64.egg
Extracting torch_scatter-1.3.1-py3.7-linux-x86_64.egg to /scratch/jojo/anaconda3/envs/torch-env/lib/python3.7/site-packages
Adding torch-scatter 1.3.1 to easy-install.pth file

Installed /scratch/jojo/anaconda3/envs/torch-env/lib/python3.7/site-packages/torch_scatter-1.3.1-py3.7-linux-x86_64.egg
Processing dependencies for torch-scatter==1.3.1
Finished processing dependencies for torch-scatter==1.3.1`

What I have tried:
1. Follow [installation guide](https://pytorch-geometric.readthedocs.io/en/latest/notes/installation.html), install with pip
2. pip uninstall and then clean pip cache, install from source: python setup.py install
3. Change gcc to 9.1.0 version, then rm -rf build && python setup.py install
4. Reinstall pytorch with gcc 6.3.0, then rm -rf build && python setup.py install
5. Clean everything, create a new environment with everything reinstalled (with gcc 6.3.0), git clone and then python setup.py install
6. Reinstall anaconda3, re-download package, install pytorch 1.2.0, build from source with gcc 6.3.0

Any suggestions?
Many thanks!

**Comment by HelloJocelynLu** [comment]
Everytime before I reinstall, I checked with 'conda list' and 'pip list' to make sure old ones have been removed.

**Comment by rusty1s** [comment]
It seems that the module skips the compilation of the CUDA files. Does you use a PyTorch installation without CUDA support? If not, it may help to set the CUDA_HOME
environment variable.

**Comment by HelloJocelynLu** [comment]
> It seems that the module skips the compilation of the CUDA files. Does you use a PyTorch installation without CUDA support? If not, it may help to set the CUDA_HOME
> environment variable.

I used Pytorch with CUDA support. 
echo $CUDA_HOME will return nothing.

But finally I get it works..
In case anyone who encounters the same problem:
(Use gcc 6.3.0 and cuda 9.2.88 for the following)
1. Using virtualenv(instead of conda, but both should work?)
2. (for pytorch) Download the [whl](https://download.pytorch.org/whl/cu92/torch_stable.html) file via pip (torch version 1.1.0; torchvision version 0.3.0)
3. Follow instructions:
pip install --verbose --no-cache-dir torch-scatter
pip install --verbose --no-cache-dir torch-sparse
pip install --verbose --no-cache-dir torch-cluster
pip install --verbose --no-cache-dir torch-spline-conv 
pip install torch-geometric

Not sure what is happening here... But I'm glad it has been solved! Thank you!
